utils/bounded_workers.go

```diff
@@ -0,0 +1,81 @@
+// (c) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package utils
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+type BoundedWorkers struct {
+	workerSpawner      chan struct{}
+	workerCount        atomic.Int32
+	outstandingWorkers sync.WaitGroup
+
+	outstandingWork sync.WaitGroup
+
+	work      chan func()
+	workClose sync.Once
+}
+
+// NewBoundedWorkers returns an instance of [BoundedWorkers] that
+// will spawn up to [max] goroutines.
+func NewBoundedWorkers(max int) *BoundedWorkers {
+	return &BoundedWorkers{
+		work:          make(chan func()),
+		workerSpawner: make(chan struct{}, max),
+	}
+}
+
+// startWorker creates a new goroutine to execute [f] immediately and then keeps the goroutine
+// alive to continue executing new work.
+func (b *BoundedWorkers) startWorker(f func()) {
+	b.workerCount.Add(1)
+	b.outstandingWorkers.Add(1)
+
+	go func() {
+		defer b.outstandingWorkers.Done()
+
+		if f != nil {
+			f()
+			b.outstandingWork.Done()
+		}
+		for f := range b.work {
+			f()
+			b.outstandingWork.Done()
+		}
+	}()
+}
+
+// Execute the given function on an existing goroutine waiting for more work, a new goroutine,
+// or return if the context is canceled.
+//
+// Execute must not be called after Wait.
+func (b *BoundedWorkers) Execute(ctx context.Context, f func()) bool {
+	b.outstandingWork.Add(1)
+	select {
+	case b.work <- f: // Feed hungry workers first.
+		return true
+	case b.workerSpawner <- struct{}{}: // Allocate a new worker to execute immediately next.
+		b.startWorker(f)
```

Comment:
Check `ctx.Done` here before starting another worker (seems possible that could hit this if we have less than `maxConcurrency)?

Reply:
Makes sense to probably put this at the top of the function first?

Reply:
We assume this will only be called if Wait hasn't been called first.